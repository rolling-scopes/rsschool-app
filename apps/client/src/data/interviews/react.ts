import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview ',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'React DOM',
      questions: [
        { id: 1011, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1012, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1013, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1020,
      name: 'Render',
      questions: [
        { id: 1021, name: 'When is a component rendered?' },
        { id: 1022, name: 'How not to render on props change?' },
        { id: 1023, name: 'Is it OK to use arrow functions in render methods?' },
      ],
    },
    {
      id: 1030,
      name: 'Interaction between components',
      questions: [
        { id: 1031, name: 'How do you pass a value from parent to child?' },
        { id: 1032, name: 'How do you pass a value from child to parent?' },
        { id: 1033, name: 'What is prop drilling?' },
        { id: 1034, name: 'Can a child component modify its own props?' },
        { id: 1035, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1040,
      name: 'Lifecycle and State',
      questions: [
        { id: 1041, name: 'What is the difference between props and state?' },
        { id: 1042, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1043, name: 'What is the component lifecycle?' },
        { id: 1044, name: 'How do you update lifecycle in function components?' },
      ],
    },
    {
      id: 1050,
      name: 'Ref',
      questions: [
        { id: 1051, name: 'What is the difference between refs and state variables?' },
        { id: 1052, name: 'When is the best time to use refs?' },
        { id: 1053, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1060,
      name: 'Context',
      questions: [
        { id: 1061, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1062, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1070,
      name: 'Redux',
      questions: [
        { id: 1071, name: 'Enumerate base principles' },
        { id: 1072, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1073, name: 'Benefits of Redux?' },
      ],
    },
    {
      id: 1080,
      name: 'Other',
      questions: [
        { id: 1081, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1082, name: 'What are the limitations of React?' },
        { id: 1083, name: 'What is a higher order component?' },
        { id: 1084, name: 'What are uncontrolled and controlled components?' },
        { id: 1085, name: 'React optimizations' },
      ],
    },
    {
      id: 1100,
      name: 'Coding task',
      questions: [{ id: 1101, name: 'Small react app: form, button, results list' }],
    },
  ],
};
