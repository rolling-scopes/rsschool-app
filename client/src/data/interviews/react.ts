import { InterviewTemplate } from './types';

export const reactTemplate: InterviewTemplate = {
  name: 'React interview',
  examplesUrl: 'https://github.com/rolling-scopes-school/tasks/blob/master/react/questions.md',
  categories: [
    {
      id: 1000,
      name: 'JSX',
      questions: [
        { id: 1001, name: 'What is JSX?' },
        { id: 1002, name: 'Is it possible to use React without JSX?' },
      ],
    },
    {
      id: 1010,
      name: 'Styled-components',
      questions: [{ id: 1011, name: 'Using styled-components' }],
    },
    {
      id: 1020,
      name: 'Components',
      questions: [
        { id: 1021, name: 'Component Definition (Class-based, Functional)' },
        { id: 1022, name: 'Components Composition' },
      ],
    },
    {
      id: 1030,
      name: 'React DOM',
      questions: [
        { id: 1031, name: 'What is the virtual DOM? How does react use the virtual DOM to render the UI?' },
        { id: 1032, name: 'Is the virtual DOM the same as the shadow DOM?' },
        { id: 1033, name: 'What is the difference between the virtual DOM and the real DOM?' },
      ],
    },
    {
      id: 1040,
      name: 'Render',
      questions: [
        { id: 1041, name: 'When is a component rendered?' },
        { id: 1042, name: 'How not to render on props change?' },
        { id: 1043, name: 'Is it OK to use arrow functions in render methods?' },
        { id: 1044, name: 'Components rendering system' },
      ],
    },
    {
      id: 1050,
      name: 'Interaction between components',
      questions: [
        { id: 1051, name: 'How do you pass a value from parent to child?' },
        { id: 1052, name: 'How do you pass a value from child to parent?' },
        { id: 1053, name: 'What is prop drilling?' },
        { id: 1054, name: 'Can a child component modify its own props?' },
        { id: 1055, name: 'How do you pass a value from sibling to sibling?' },
      ],
    },
    {
      id: 1060,
      name: 'Lifecycle and State',
      questions: [
        { id: 1061, name: 'What is the difference between props and state?' },
        { id: 1062, name: 'How does state in a class component differ from state in a functional component?' },
        { id: 1063, name: 'What is the component lifecycle?' },
        { id: 1064, name: 'How do you update lifecycle in function components?' },
        { id: 1065, name: 'Controlled/uncontrolled components' },
        { id: 1066, name: 'Stateful vs stateless components' },
      ],
    },
    {
      id: 1070,
      name: 'Ref',
      questions: [
        { id: 1071, name: 'What is the difference between refs and state variables?' },
        { id: 1072, name: 'When is the best time to use refs?' },
        { id: 1073, name: 'What is the proper way to update a ref in a function component?' },
      ],
    },
    {
      id: 1080,
      name: 'Context',
      questions: [
        { id: 1081, name: 'What is the difference between the context API and prop drilling?' },
        { id: 1082, name: "When shouldn't you use the context API?" },
      ],
    },
    {
      id: 1090,
      name: 'Redux',
      questions: [
        { id: 1091, name: 'Enumerate base principles' },
        { id: 1092, name: 'What is the typical flow of data in a React + Redux app?' },
        { id: 1093, name: 'Benefits of Redux?' },
        { id: 1094, name: 'Async Redux flow' },
      ],
    },
    {
      id: 1100,
      name: 'State Management by managers',
      questions: [
        { id: 1101, name: 'Mobx' },
        { id: 1102, name: 'Reflux' },
      ],
    },
    {
      id: 1110,
      name: 'Routing',
      questions: [
        { id: 1111, name: 'React Router' },
        { id: 1112, name: 'History' },
      ],
    },
    {
      id: 1120,
      name: 'Other',
      questions: [
        { id: 1121, name: 'Is it a good idea to use Math.random for keys?' },
        { id: 1122, name: 'What are the limitations of React?' },
        { id: 1123, name: 'What is a higher order component?' },
        { id: 1124, name: 'What are uncontrolled and controlled components?' },
        { id: 1125, name: 'React optimizations' },
      ],
    },
    {
      id: 1130,
      name: 'Coding task',
      questions: [{ id: 1131, name: 'Small react app: form, button, results list' }],
    },
    {
      id: 1140,
      name: 'Next.js',
      questions: [
        { id: 1141, name: 'What is Next.js, and how does it differ from traditional React applications?' },
        { id: 1142, name: 'Explain the concept of server-side rendering (SSR) in Next.js' },
        { id: 1143, name: 'How does Next.js handle client-side routing?' },
        { id: 1144, name: 'What are the benefits of using Next.js for building React applications?' },
        {
          id: 1145,
          name: 'What is the purpose of the getInitialProps function in Next.js? How is it different from using getStaticProps or getServerSideProps?',
        },
        { id: 1146, name: 'How does Next.js handle automatic code splitting, and why is it important?' },
        {
          id: 1147,
          name: 'What is the purpose of the Link component in Next.js, and how does it differ from traditional anchor (<a>) tags?',
        },
        { id: 1148, name: 'How can you configure routing in a Next.js application?' },
      ],
    },
    {
      id: 1150,
      name: 'Data Fetching',
      questions: [
        {
          id: 1151,
          name: 'Compare and contrast getStaticProps and getServerSideProps in terms of use cases and performance considerations',
        },
        {
          id: 1152,
          name: 'When would you choose to use getStaticPaths in Next.js, and how does it relate to dynamic routes?',
        },
      ],
    },
    {
      id: 1160,
      name: 'Advanced Topics',
      questions: [
        { id: 1161, name: 'Describe the purpose and use cases of API routes in Next.js' },
        { id: 1162, name: 'Explain the concept of middleware in Next.js and its role in the application lifecycle' },
        { id: 1163, name: 'How does Next.js handle authentication in applications?' },
      ],
    },
  ],
};
