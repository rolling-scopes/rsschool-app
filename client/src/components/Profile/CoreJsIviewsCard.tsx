import * as React from 'react';
import isEqual from 'lodash/isEqual';
import { Typography, List, Button } from 'antd';
import CommonCard from './CommonCard';
import CoreJsIviewsModal from './CoreJsIviewsModal';

const { Text } = Typography;

import { QuestionCircleOutlined, FullscreenOutlined } from '@ant-design/icons';

export interface CoreJsInterviewData {
  courseFullName: string;
  courseName: string;
  locationName: string;
  interview: {
    answers: {
      answer: string;
      questionText: string;
      questionId: string;
    }[];
    interviewer: {
      name: string;
      githubId: string;
    };
    comment: string;
    score: number;
  };
}

type Props = {
  data: CoreJsInterviewData[];
};

type State = {
  courseIndex: number;
  isCoreJsIviewModalVisible: boolean;
};

class CoreJSIviewsCard extends React.Component<Props, State> {
  state = {
    courseIndex: 0,
    isCoreJsIviewModalVisible: false,
  };

  shouldComponentUpdate = (_: Props, nextState: State) =>
    !isEqual(nextState.isCoreJsIviewModalVisible, this.state.isCoreJsIviewModalVisible);

  private showCoreJsIviewModal = (courseIndex: number) => {
    this.setState({ courseIndex, isCoreJsIviewModalVisible: true });
  };

  private hideCoreJsIviewModal = () => {
    this.setState({ isCoreJsIviewModalVisible: false });
  };

  render() {
    const stats = this.props.data;
    const { isCoreJsIviewModalVisible } = this.state;
    const { courseIndex } = this.state;

    return (
      <>
        <CoreJsIviewsModal
          stats={stats[courseIndex]}
          isVisible={isCoreJsIviewModalVisible}
          onHide={this.hideCoreJsIviewModal}
        />
        <CommonCard
          title="CoreJS Interviews"
          icon={<QuestionCircleOutlined />}
          content={
            <List
              itemLayout="horizontal"
              dataSource={stats}
              renderItem={({ courseName, locationName, interview: { score, interviewer } }, idx) => (
                <List.Item style={{ display: 'flex', justifyContent: 'space-between' }}>
                  <div style={{ flexGrow: 2 }}>
                    <p style={{ marginBottom: 5 }}>
                      <Text strong>
                        {courseName}
                        {locationName && ` / ${locationName}`}
                      </Text>
                    </p>
                    {
                      <p style={{ fontSize: 12, marginBottom: 5 }}>
                        Score: <Text mark>{score}</Text>
                      </p>
                    }
                    {
                      <p style={{ fontSize: 12, marginBottom: 5 }}>
                        Interviewer: <a href={`/profile?githubId=${interviewer.githubId}`}>{interviewer.name}</a>
                      </p>
                    }
                  </div>
                  <Button type="dashed" onClick={this.showCoreJsIviewModal.bind(null, idx)}>
                    <FullscreenOutlined />
                  </Button>
                </List.Item>
              )}
            />
          }
        />
      </>
    );
  }
}

export default CoreJSIviewsCard;
