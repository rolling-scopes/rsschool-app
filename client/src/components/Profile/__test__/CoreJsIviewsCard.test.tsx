import React from 'react';
import { act, render, screen } from '@testing-library/react';
import CoreJsIviewsCard, { CoreJsInterviewsData } from '../CoreJsIviewsCard';

describe('CoreJSIviewsCard', () => {
  const data = [
    {
      courseFullName: 'rs-2019',
      courseName: 'rs-2019',
      locationName: 'minsk',
      interviews: [
        {
          answers: [
            {
              answer: true,
              questionText: 'test',
              questionId: 'test1',
            },
            {
              answer: false,
              questionText: 'test',
              questionId: 'test2',
            },
          ],
          interviewer: {
            name: 'Dzmitry Petrov',
            githubId: 'dima',
          },
          comment: 'test',
          score: 4,
          name: 'CoreJS Interview',
        },
      ],
    },
  ] as CoreJsInterviewsData[];

  it('should render correctly', () => {
    const { container } = render(<CoreJsIviewsCard data={data} />);
    expect(container).toMatchSnapshot();
  });

  it('should show modal', async () => {
    render(<CoreJsIviewsCard data={data} />);

    const btn = await screen.findByTestId('profile-corejs-iview-button');

    act(() => {
      btn.click();
    });

    const modal = await screen.findByTestId('profile-corejs-iviews-modal-table');
    expect(modal).toBeInTheDocument();
  });
});
