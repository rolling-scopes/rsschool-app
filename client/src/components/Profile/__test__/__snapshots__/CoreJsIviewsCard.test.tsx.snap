// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreJSIviewsCard should render correctly 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered css-dev-only-do-not-override-nnuwmp"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <h2
            class="ant-typography ant-typography-ellipsis ant-typography-single-line ant-typography-ellipsis-single-line css-dev-only-do-not-override-nnuwmp"
            style="font-size: 16px; margin-bottom: 0px; display: flex; align-items: center; justify-content: space-between;"
          >
            <span>
              <span
                aria-label="question-circle"
                class="anticon anticon-question-circle"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="question-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                  />
                  <path
                    d="M623.6 316.7C593.6 290.4 554 276 512 276s-81.6 14.5-111.6 40.7C369.2 344 352 380.7 352 420v7.6c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8V420c0-44.1 43.1-80 96-80s96 35.9 96 80c0 31.1-22 59.6-56.1 72.7-21.2 8.1-39.2 22.3-52.1 40.9-13.1 19-19.9 41.8-19.9 64.9V620c0 4.4 3.6 8 8 8h48c4.4 0 8-3.6 8-8v-22.7a48.3 48.3 0 0130.9-44.8c59-22.7 97.1-74.7 97.1-132.5.1-39.3-17.1-76-48.3-103.3zM472 732a40 40 0 1080 0 40 40 0 10-80 0z"
                  />
                </svg>
              </span>
               
              CoreJS Interviews
            </span>
          </h2>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
    >
      <div
        class="ant-list ant-list-split css-dev-only-do-not-override-nnuwmp"
      >
        <div
          class="ant-spin-nested-loading css-dev-only-do-not-override-nnuwmp"
        >
          <div
            class="ant-spin-container"
          >
            <ul
              class="ant-list-items"
            >
              <li
                class="ant-list-item"
                style="display: flex; justify-content: space-between;"
              >
                <div
                  style="flex-grow: 2;"
                >
                  <p
                    style="margin-bottom: 5px;"
                  >
                    <span
                      class="ant-typography css-dev-only-do-not-override-nnuwmp"
                    >
                      <strong>
                        rs-2019
                         / minsk
                      </strong>
                    </span>
                  </p>
                  <p
                    style="margin-bottom: 5px;"
                  >
                    CoreJS Interview
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Score: 
                    <span
                      class="ant-typography css-dev-only-do-not-override-nnuwmp"
                    >
                      <mark>
                        4
                      </mark>
                    </span>
                  </p>
                  <p
                    style="font-size: 12px; margin-bottom: 5px;"
                  >
                    Interviewer: 
                    <a
                      href="/profile?githubId=dima"
                    >
                      Dzmitry Petrov
                    </a>
                  </p>
                </div>
                <button
                  class="ant-btn css-dev-only-do-not-override-nnuwmp ant-btn-dashed"
                  data-testid="profile-corejs-iview-button"
                  type="button"
                >
                  <span
                    aria-label="fullscreen"
                    class="anticon anticon-fullscreen"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="fullscreen"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M290 236.4l43.9-43.9a8.01 8.01 0 00-4.7-13.6L169 160c-5.1-.6-9.5 3.7-8.9 8.9L179 329.1c.8 6.6 8.9 9.4 13.6 4.7l43.7-43.7L370 423.7c3.1 3.1 8.2 3.1 11.3 0l42.4-42.3c3.1-3.1 3.1-8.2 0-11.3L290 236.4zm352.7 187.3c3.1 3.1 8.2 3.1 11.3 0l133.7-133.6 43.7 43.7a8.01 8.01 0 0013.6-4.7L863.9 169c.6-5.1-3.7-9.5-8.9-8.9L694.8 179c-6.6.8-9.4 8.9-4.7 13.6l43.9 43.9L600.3 370a8.03 8.03 0 000 11.3l42.4 42.4zM845 694.9c-.8-6.6-8.9-9.4-13.6-4.7l-43.7 43.7L654 600.3a8.03 8.03 0 00-11.3 0l-42.4 42.3a8.03 8.03 0 000 11.3L734 787.6l-43.9 43.9a8.01 8.01 0 004.7 13.6L855 864c5.1.6 9.5-3.7 8.9-8.9L845 694.9zm-463.7-94.6a8.03 8.03 0 00-11.3 0L236.3 733.9l-43.7-43.7a8.01 8.01 0 00-13.6 4.7L160.1 855c-.6 5.1 3.7 9.5 8.9 8.9L329.2 845c6.6-.8 9.4-8.9 4.7-13.6L290 787.6 423.7 654c3.1-3.1 3.1-8.2 0-11.3l-42.4-42.4z"
                      />
                    </svg>
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
